src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                    compilationContext, GetDefinitionLocationToFade, SupportsDiscard,
+                    UnusedValueExpressionStatementOption, UnusedValueAssignmentOption));
+
+        private sealed class SymbolStartAnalyzer
+        {
+            private readonly Func<IOperation, Location> _getDefinitionLocationToFade;
+            private readonly Func<SyntaxTree, bool> _supportsDiscard;
+            private readonly Option<CodeStyleOption<UnusedValuePreference>> _unusedValueExpressionStatementOption;
+            private readonly Option<CodeStyleOption<UnusedValuePreference>> _unusedValueAssignmentOption;
+
+            private readonly INamedTypeSymbol _eventArgsType, _interlockedType, _immutableInterlockedType;
+            private readonly ImmutableHashSet<INamedTypeSymbol> _attributeSetForMethodsToIgnore;
+            private readonly ConcurrentDictionary<IParameterSymbol, bool> _unusedParameters;
+            private readonly ConcurrentDictionary<IMethodSymbol, bool> _methodsUsedAsDelegates;
+
+            public SymbolStartAnalyzer(
+                Func<IOperation, Location> getDefinitionLocationToFade,
+                Func<SyntaxTree, bool> supportsDiscard,
+                Option<CodeStyleOption<UnusedValuePreference>> unusedValueExpressionStatementOption,
+                Option<CodeStyleOption<UnusedValuePreference>> unusedValueAssignmentOption,
+                INamedTypeSymbol eventArgsType,
+                INamedTypeSymbol interlockedType,
+                INamedTypeSymbol immutableInterlockedType,
+                ImmutableHashSet<INamedTypeSymbol> attributeSetForMethodsToIgnore)
+            {
+                _getDefinitionLocationToFade = getDefinitionLocationToFade;
+                _supportsDiscard = supportsDiscard;
+                _unusedValueExpressionStatementOption = unusedValueExpressionStatementOption;
+                _unusedValueAssignmentOption = unusedValueAssignmentOption;
+
+                _eventArgsType = eventArgsType;
+                _interlockedType = interlockedType;
+                _immutableInterlockedType = immutableInterlockedType;
+                _attributeSetForMethodsToIgnore = attributeSetForMethodsToIgnore;
+                _unusedParameters = new ConcurrentDictionary<IParameterSymbol, bool>();
+                _methodsUsedAsDelegates = new ConcurrentDictionary<IMethodSymbol, bool>();
+            }
+
+            public static void CreateAndRegisterActions(
+                CompilationStartAnalysisContext context,
+                Func<IOperation, Location> getDefinitionLocationToFade,
+                Func<SyntaxTree, bool> supportsDiscard,
+                Option<CodeStyleOption<UnusedValuePreference>> unusedValueExpressionStatementOption,
+                Option<CodeStyleOption<UnusedValuePreference>> unusedValueAssignmentOption)
+            {
+                var attributeSetForMethodsToIgnore = ImmutableHashSet.CreateRange(GetAttributesForMethodsToIgnore(context.Compilation));
```

Comment:
it feels like this should be user extensible.  Do we have an editorconfig option that allows users to specify a comma-delimited set of attribute names to ignore?  That would be a good escape hatch i think.

Reply:
This matches the commonly used attributes that were used by FxCop rule to bail out on for unused parameter diagnostic. I agree we can make this extensible, but I am going to delegate that to a separate tracking work item.

---
In reply to: [230604812](https://github.com/dotnet/roslyn/pull/30777#discussion_r230604812) [](ancestors = 230604812)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/31008](https://github.com/dotnet/roslyn/issues/31008)

---
In reply to: [230894780](https://github.com/dotnet/roslyn/pull/30777#discussion_r230894780) [](ancestors = 230894780,230604812)